src/util/trace.h

```diff
@@ -7,37 +7,26 @@
 
 #ifdef ENABLE_TRACING
 
+// Setting SDT_USE_VARIADIC let's systemtap (sys/sdt.h) know that we want to use
+// the optional variadic macros to define tracepoints.
+#define SDT_USE_VARIADIC 1
+
 #include <sys/sdt.h>
 
-#define TRACE(context, event) DTRACE_PROBE(context, event)
-#define TRACE1(context, event, a) DTRACE_PROBE1(context, event, a)
-#define TRACE2(context, event, a, b) DTRACE_PROBE2(context, event, a, b)
-#define TRACE3(context, event, a, b, c) DTRACE_PROBE3(context, event, a, b, c)
-#define TRACE4(context, event, a, b, c, d) DTRACE_PROBE4(context, event, a, b, c, d)
-#define TRACE5(context, event, a, b, c, d, e) DTRACE_PROBE5(context, event, a, b, c, d, e)
-#define TRACE6(context, event, a, b, c, d, e, f) DTRACE_PROBE6(context, event, a, b, c, d, e, f)
-#define TRACE7(context, event, a, b, c, d, e, f, g) DTRACE_PROBE7(context, event, a, b, c, d, e, f, g)
-#define TRACE8(context, event, a, b, c, d, e, f, g, h) DTRACE_PROBE8(context, event, a, b, c, d, e, f, g, h)
-#define TRACE9(context, event, a, b, c, d, e, f, g, h, i) DTRACE_PROBE9(context, event, a, b, c, d, e, f, g, h, i)
-#define TRACE10(context, event, a, b, c, d, e, f, g, h, i, j) DTRACE_PROBE10(context, event, a, b, c, d, e, f, g, h, i, j)
-#define TRACE11(context, event, a, b, c, d, e, f, g, h, i, j, k) DTRACE_PROBE11(context, event, a, b, c, d, e, f, g, h, i, j, k)
-#define TRACE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l) DTRACE_PROBE12(context, event, a, b, c, d, e, f, g, h, i, j, k, l)
+// A USDT tracepoint with no arguments.
+#define TRACEPOINT0(context, event) \
```

Comment:
Macro magic to the rescue! This works for me:

```patch
diff --git a/src/test/util_trace_tests.cpp b/src/test/util_trace_tests.cpp
index 6e3e846b4f..952b9cc6e4 100644
--- a/src/test/util_trace_tests.cpp
+++ b/src/test/util_trace_tests.cpp
@@ -18,7 +18,7 @@ BOOST_FIXTURE_TEST_SUITE(util_trace_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(test_tracepoint_zero_args)
 {
-  TRACEPOINT0(test, zero_args);
+  TRACEPOINT(test, zero_args);
   BOOST_CHECK(true);
 }
 
diff --git a/src/util/trace.h b/src/util/trace.h
index 816d8c295d..3a0aed391b 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -29,25 +29,19 @@
     unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
 
 // Returns true if something is attached to the tracepoint.
-#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
-
-// A USDT tracepoint with no arguments.
-#define TRACEPOINT0(context, event) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBE(context, event)
+#define TRACEPOINT_ACTIVE(context, event, ...) context##_##event##_semaphore > 0
 
 // A USDT tracepoint with one to twelve arguments. It's checked that the
 // tracepoint is active before preparing its arguments.
-#define TRACEPOINT(context, event, ...) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBEV(context, event, __VA_ARGS__)
+#define TRACEPOINT(context, ...) \
+    if (TRACEPOINT_ACTIVE(context, __VA_ARGS__)) \
+        STAP_PROBEV(context, __VA_ARGS__)
 
 #else
 
 #define TRACEPOINT_SEMAPHORE(context, event)
 #define TRACEPOINT_ACTIVE(context, event) false
-#define TRACEPOINT0(context, event)
-#define TRACEPOINT(context, event, ...)
+#define TRACEPOINT(context, ...)
 
 #endif // ENABLE_TRACING
 ```

The trick is to make `event` part of the variadic macro, so that there's always at least 1 argument in the variadic

Reply:
Based on @willcl-ark's comment https://github.com/bitcoin/bitcoin/issues/26916#issuecomment-1480997053 here is a version that works without any warnings in clang and gcc:

```patch
diff --git a/src/test/util_trace_tests.cpp b/src/test/util_trace_tests.cpp
index 6e3e846b4f..952b9cc6e4 100644
--- a/src/test/util_trace_tests.cpp
+++ b/src/test/util_trace_tests.cpp
@@ -18,7 +18,7 @@ BOOST_FIXTURE_TEST_SUITE(util_trace_tests, BasicTestingSetup)
 
 BOOST_AUTO_TEST_CASE(test_tracepoint_zero_args)
 {
-  TRACEPOINT0(test, zero_args);
+  TRACEPOINT(test, zero_args);
   BOOST_CHECK(true);
 }
 
diff --git a/src/util/trace.h b/src/util/trace.h
index 816d8c295d..aa88771b1e 100644
--- a/src/util/trace.h
+++ b/src/util/trace.h
@@ -26,28 +26,41 @@
 // upon attaching to the tracepoint and decremented when detaching. This needs
 // to be a global variable. It's placed in the '.probes' ELF section.
 #define TRACEPOINT_SEMAPHORE(context, event) \
-    unsigned short context##_##event##_semaphore __attribute__((section (".probes")))
+    unsigned short context##_##event##_semaphore __attribute__((section(".probes")))
+
+// Extract the first argument of a variable number of arguments, even without warning
+// when only 1 argument is provided
+#define TRACEPOINT_FIRST_ARG(...) TRACEPOINT_FIRST_ARG_HELPER(__VA_ARGS__, dummy)
+#define TRACEPOINT_FIRST_ARG_HELPER(arg1, ...) arg1
 
 // Returns true if something is attached to the tracepoint.
-#define TRACEPOINT_ACTIVE(context, event) context##_##event##_semaphore > 0
+#define TRACEPOINT_ACTIVE(context, event) TRACEPOINT_ACTIVE_HELPER(context, event)
+#define TRACEPOINT_ACTIVE_HELPER(context, event) context##_##event##_semaphore > 0
 
-// A USDT tracepoint with no arguments.
-#define TRACEPOINT0(context, event) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBE(context, event)
+#if defined(__clang__)
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH _Pragma("clang diagnostic push") _Pragma("clang diagnostic ignored \"-Wgnu-zero-variadic-macro-arguments\"")
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP _Pragma("clang diagnostic pop")
+#else
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH
+#define TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP
+#endif
 
 // A USDT tracepoint with one to twelve arguments. It's checked that the
 // tracepoint is active before preparing its arguments.
-#define TRACEPOINT(context, event, ...) \
-    if (TRACEPOINT_ACTIVE(context, event)) \
-        STAP_PROBEV(context, event, __VA_ARGS__)
+#define TRACEPOINT(context, ...)                                             \
+    do {                                                                     \
+        if (TRACEPOINT_ACTIVE(context, TRACEPOINT_FIRST_ARG(__VA_ARGS__))) { \
+            TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_PUSH                       \
+            STAP_PROBEV(context, __VA_ARGS__);                               \
+            TRACEPOINT_DISABLE_WARN_ZERO_VARIADIC_POP                        \
+        }                                                                    \
+    } while (0)
 
 #else
 
 #define TRACEPOINT_SEMAPHORE(context, event)
 #define TRACEPOINT_ACTIVE(context, event) false
-#define TRACEPOINT0(context, event)
-#define TRACEPOINT(context, event, ...)
+#define TRACEPOINT(context, ...)
 
 #endif // ENABLE_TRACING
```

Reply:
Do you remember why you added a `do { ... } while (0)` to the `TRACEPOINT` macro?